File: TypeScript/NumberOfGoodPairs.ts
// HASHMAP
// https://leetcode.com/problems/number-of-good-pairs/

// time complexity: o(n)
// space complexity: o(n)

function numIdenticalPairs(nums: number[]): number {
    const map: Map<number, number> = new Map<number, number>()
    let sum = 0

    for(let i=0; i < nums.length; i++) {
        if(map.has(nums[i])) {
            sum += map.get(nums[i]) || 0 //factorial ??? study behavior
            map.set(nums[i], (map.get(nums[i]) || 0) + 1)
        } else {
            map.set(nums[i], 1)
        }
    }

    return sum
};


console.log(numIdenticalPairs([1,2, 2, 3, 2, 2])) // -> 6
console.log(numIdenticalPairs([1,2, 2, 3,])) // -> 1
console.log(numIdenticalPairs([1,2,3])) // -> 0

File: Go/NumberOfGoodPairs.go
package Go

// HASHMAP
// https://leetcode.com/problems/number-of-good-pairs/

// time complexity: o(n)
// space complexity: o(n)

// Math way :)
func numberOfGoodPairsMath(nums []int) int {
	counter := make(map[int]int)
	var result int

	for _, num := range nums {
		counter[num]++
	}

	for _, value := range counter {
		result += (value * (value - 1) / 2)
	}

	return result
}

// Math haters way :)
func numberOfGoodPairsNotMath(nums []int) int {
	counter := make(map[int]int)
	var result int

	for _, num := range nums {
		result += counter[num]
		counter[num]++
	}

	return result
}
